### PR TITLE
Ddp 5762 refactor hide show blocks

### PR DESCRIPTION
**Pull request by MocanaAtBroad**
**Needs to be re-reviewed after merge and minor refactoring of imports.**
Trying to deal with a few issues:
1. Centralizing the hide/show functionality on ActivitySection (before it was done higher up in component hierarchy, but hide/show is done at the level of Blocks and these are below the Section level. This is important because ActivitySection component is also reused by the embedded and modal activity blocks. This is main reason functionality was not working in these blocks. The key change here: https://github.com/broadinstitute/ddp-angular/pull/500/files#diff-45705fe892c0b18587e0c4c566460622e324b44d968c0e6b64c9c5f069b641a2R122
2. To centralize the hide/show we implement a recursive collection of all the relevant blocks (the model for blocks is all subclassed from ActivityBlock. https://github.com/broadinstitute/ddp-angular/pull/500/files#r615435297 . All the blocks come together in this short little snippet:  https://github.com/broadinstitute/ddp-angular/pull/500/files#diff-d7a01d134bc92c1387f6103f133dd9dd025c669bee24befe6df5da24ed92760bR20
2. Got rid of the visiblityChanged output parameters. They were not being used as far as I can see. If someone can find a use them do let me know and will bring them back.

### CODE REVIEW

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts [hunk 1]
```diff
@@ -93,22 +94,44 @@ import { ActivityActivityBlock } from '../../models/activity/activityActivityBlo
...
+
+    ngOnDestroy(): void {
+        this.subscription.unsubscribe();
     }
 
     public updateVisibility(visibility: BlockVisibility[]): void {
-        this.visibilityChanged.emit(visibility);
+        let blockVisibilityChanged = false;
+        visibility.forEach(element => {
+            this.section.allChildBlocks().forEach(block => {
```
Comment (alextanlan) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts:125]: Nice idea!

**Review by alextanlan — DISMISSED** [review]

File: ddp-workspace/projects/ddp-sdk/src/lib/models/activity/activityBlock.ts [hunk 2]
```diff
@@ -8,6 +8,11 @@ export abstract class ActivityBlock {
 
     abstract get blockType(): BlockType;
 
+    /**
+     * all block contained by this block, including self
+     */
+    abstract get blocks(): Array<ActivityBlock>;
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/models/activity/activityBlock.ts:14]: Force every implementation to return blocks under it's control, including itself. Sets up a way to recurse through all the blocks.

File: ddp-workspace/projects/ddp-sdk/src/lib/models/activity/conditionalBlock.ts [hunk 3]
```diff
@@ -12,6 +12,12 @@ export class ConditionalBlock extends ActivityBlock {
         return BlockType.Conditional;
     }
 
+    public get blocks(): Array<ActivityBlock> {
+        return [this as ActivityBlock]
+            .concat(this.controlQuestion.blocks as ActivityBlock[])
+            .concat(this.nestedGroupBlock.blocks as ActivityBlock[]);
+    }
+
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/models/activity/conditionalBlock.ts:20]: One of the more interesting implementations of this method.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/conditionalBlock.component.ts [hunk 4]
```diff
@@ -24,39 +23,15 @@ import { CompositeDisposable } from '../../../compositeDisposable';
...
-
-    public updateVisibility(visibility: BlockVisibility[]): void {
-        visibility.forEach(shownBlockStatus => {
-            const matchingBlock = this.block.nestedGroupBlock
-                .nestedBlocks.find(nestedBlock => nestedBlock.id === shownBlockStatus.blockGuid);
-            if (matchingBlock) {
-                matchingBlock.shown = shownBlockStatus.shown;
-            }
-        });
-    }
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activity-blocks/conditionalBlock.component.ts:61]: ConditionalBlocks supposedly exist to control hide/show. It does provide structure to set this up, but reality is that that server can tell any block to hide or show any time. So all this code here is not necessary.

File: ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts [hunk 5]
```diff
@@ -93,22 +94,44 @@ import { ActivityActivityBlock } from '../../models/activity/activityActivityBlo
...
+                if (block.id === element.blockGuid) {
+                    if (block.shown !== element.shown) {
+                        block.shown = element.shown;
+                        blockVisibilityChanged = true;
+                    }
+                }
+            });
+        });
+        blockVisibilityChanged && this.cdr.detectChanges();
```
Comment (MocanaAtBroad) [inline ddp-workspace/projects/ddp-sdk/src/lib/components/activityForm/activitySection.component.ts:134]: No need to trigger detectChanges() unless we know something has changed.

**Review by aglushko-broad — APPROVED** [review]

**Pull request merged**